pkg/3scale/amp/operator/system_reconciler.go

```diff
@@ -182,6 +143,134 @@ func (r *SystemReconciler) Reconcile() (reconcile.Result, error) {
+	}
+	if generationChanged {
+		err = helper.DeleteJob(component.SystemAppPostHookJobName, r.apiManager.GetNamespace(), r.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Used to synchronize the system-app Deployment with the PreHook/PostHook Jobs
+	currentAppDeploymentGeneration, err := getSystemAppDeploymentGeneration(r.apiManager.GetNamespace(), r.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// SystemApp PreHook Job
+	preHookJob := system.AppPreHookJob(ampImages.Options.SystemImage, currentAppDeploymentGeneration)
+	err = r.ReconcileJob(preHookJob, reconcilers.CreateOnlyMutator)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Block reconciling system-app Deployment until PreHook Job has completed
+	if !helper.HasJobCompleted(preHookJob.Name, preHookJob.Namespace, r.Client()) {
+		systemComponentNotReady = true
+	}
+
+	// SystemApp Deployment
+	systemAppDeploymentMutators := []reconcilers.DMutateFn{
+		reconcilers.DeploymentAffinityMutator,
+		reconcilers.DeploymentTolerationsMutator,
+		reconcilers.DeploymentPodTemplateLabelsMutator,
+		reconcilers.DeploymentPriorityClassMutator,
+		reconcilers.DeploymentTopologySpreadConstraintsMutator,
+		reconcilers.DeploymentPodTemplateAnnotationsMutator,
+		r.systemAppDeploymentResourceMutator,
+		reconcilers.DeploymentRemoveDuplicateEnvVarMutator,
+		reconcilers.DeploymentPodContainerImageMutator,
+	}
+	if r.apiManager.Spec.System.AppSpec.Replicas != nil {
+		systemAppDeploymentMutators = append(systemAppDeploymentMutators, reconcilers.DeploymentReplicasMutator)
+	}
+	if !systemComponentNotReady {
+		err = r.ReconcileDeployment(system.AppDeployment(ampImages.Options.SystemImage), reconcilers.DeploymentMutator(systemAppDeploymentMutators...))
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Block reconciling PostHook Job unless BOTH the PreHook Job has completed and the system-app Deployment is ready and not in the process of updating
+	deployment := &k8sappsv1.Deployment{}
+	err = r.Client().Get(context.TODO(), k8sclient.ObjectKey{
+		Namespace: r.apiManager.GetNamespace(),
+		Name:      component.SystemAppDeploymentName,
+	}, deployment)
+	if err != nil && !k8serr.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
+	if k8serr.IsNotFound(err) || !helper.IsDeploymentAvailable(deployment) || helper.IsDeploymentProgressing(deployment) || !helper.HasJobCompleted(preHookJob.Name, preHookJob.Namespace, r.Client()) {
+		systemComponentNotReady = true
+	}
+
+	// SystemApp PostHook Job
+	if !systemComponentNotReady {
+		err = r.ReconcileJob(system.AppPostHookJob(ampImages.Options.SystemImage, currentAppDeploymentGeneration), reconcilers.CreateOnlyMutator)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// 3scale 2.14 -> 2.15
+	isMigrated, err := upgrade.MigrateDeploymentConfigToDeployment(component.SystemAppDeploymentName, r.apiManager.GetNamespace(), false, r.Client(), nil)
```

Comment:
Q
Can you describe the flow of 2.14 > 2.15 upgrade please. 
In my head:
- create job and run it
- mutators = we should get deployment at this stage
- migrate = remove the dc
- posthook?

Reply:
1. Create PreHook Job and run it
2. Once the PreHook Job has completed, create the `system-app` Deployment
3. Once the `system-app` Deployment is ready, create the PostHook Job and run it
4. Migrate the DeploymentConfig to Deployment (i.e. delete the old DeploymentConfig now that the Deployment is healthy)